rust/main/agents/relayer/src/server/message_retry.rs

```diff
@@ -1,32 +1,87 @@
-use crate::settings::matching_list::MatchingList;
+use crate::{server::ENDPOINT_MESSAGES_QUEUE_SIZE, settings::matching_list::MatchingList};
+
 use axum::{extract::State, routing, Json, Router};
+
 use derive_new::new;
-use tokio::sync::broadcast::Sender;
+use serde::{Deserialize, Serialize};
+use tokio::sync::{broadcast::Sender, mpsc};
 
 const MESSAGE_RETRY_API_BASE: &str = "/message_retry";
 
-#[derive(new, Clone)]
+#[derive(new)]
 pub struct MessageRetryApi {
-    tx: Sender<MatchingList>,
+    retry_request_transmitter: Sender<MessageRetryRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryApiState {
+    pub retry_request_transmitter: Sender<MessageRetryRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryRequest {
+    pub uuid: String,
+    pub pattern: MatchingList,
+    pub transmitter: mpsc::Sender<MessageRetryResponse>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct MessageRetryResponse {
+    /// ID of the retry request
+    pub uuid: String,
+    /// how many pending operations were evaluated
+    pub evaluated: usize,
+    /// how many of the pending operations matched the retry request pattern
+    pub matched: u64,
 }
 
 async fn retry_message(
-    State(tx): State<Sender<MatchingList>>,
+    State(state): State<MessageRetryApiState>,
     Json(retry_req_payload): Json<MatchingList>,
-) -> String {
-    match tx.send(retry_req_payload) {
-        Ok(_) => "Moved message(s) to the front of the queue".to_string(),
-        // Technically it's bad practice to print the error message to the user, but
-        // this endpoint is for debugging purposes only.
-        Err(err) => format!("Failed to send retry request to the queue: {}", err),
+) -> Result<Json<MessageRetryResponse>, String> {
+    let uuid = uuid::Uuid::new_v4();
+    let uuid_string = uuid.to_string();
+
+    tracing::debug!(uuid = uuid_string, "Sending message retry request");
+
+    let (transmitter, mut receiver) = mpsc::channel(ENDPOINT_MESSAGES_QUEUE_SIZE);
+    state
+        .retry_request_transmitter
+        .send(MessageRetryRequest {
+            uuid: uuid_string.clone(),
+            pattern: retry_req_payload,
+            transmitter,
+        })
+        .map_err(|err| {
+            // Technically it's bad practice to print the error message to the user, but
+            // this endpoint is for debugging purposes only.
+            format!("Failed to send retry request to the queue: {}", err)
+        })?;
+
+    let mut resp = MessageRetryResponse {
+        uuid: uuid_string,
+        evaluated: 0,
+        matched: 0,
+    };
+
+    // Wait for responses from relayer
+    tracing::debug!(uuid = resp.uuid, "Waiting for response from relayer");
+    while let Some(relayer_resp) = receiver.recv().await {
+        tracing::debug!(uuid = resp.uuid, "Relayer response");
+        resp.evaluated += relayer_resp.evaluated;
+        resp.matched += relayer_resp.matched;
     }
+
+    Ok(Json(resp))
```

Comment:
I think it'd make things easier to debug if we returned a map of responses by uuid in the json response

Reply:
What do you mean by this? I think currently the prepare, submit and confirm queue all return the same UUID :thinking: 

Or are you thinking of something like:

```json
{
    "prepare_queue": {
        "evaluated": ...,
        "matched": ...,
   },
    "submit_queue": {
        "evaluated": ...,
        "matched": ...,
   },
    "confirm_queue": {
        "evaluated": ...,
        "matched": ...,
   }
}
```

Reply:
oh right, every message has its own channel actually and you're just adding up responses from each sender, I missed that

Things are a bit more complicated:
- we should use `try_recv` instead of `recv` everywhere, to avoid blocking on an empty channel. However because different senders may take varying amounts of time to respond to the request and the channel may be spuriously empty before we received all responses:
- we need to wait on a response from every sender (`num_chains * num_queues`). So we should keep a counter and only return once we got responses from all `receiver.sender_strong_count()`

Reply:
your logic to add up the response fields makes sense now